lib/BaseN.js

```diff
@@ -0,0 +1,196 @@
+'use strict';
+
+// external requires
+var assert = require('assert-plus');
+
+// local const
+var DEFAULT_OPTIONS = {
+    characters:
+        '0123456789abcdefghijklmnopqrstuvwxyzABCDEFGHIJKLMNOPQRSTUVWXYZ_-'
+    //   0123456789 123456789 123456789 123456789 123456789 123456789 123
+    //             1         2         3         4         5         6
+    // comment above helps decipher encoding when debugging.
+};
+
+
+/**
+ * BaseN. Uses a baseN scheme to encode/decode integers
+ * @param {Object} options an options object
+ * @constructor
+ */
+function BaseN(options) {
+
+    var self = this;
+    var opt = options || {};
+
+    // assert
+    if (opt.max && opt.length) {
+        throw new Error('cannot specify both max and length');
+    }
+
+    if (opt.base && opt.characters) {
+        throw new Error('cannot specify both base and characters');
+    }
+
+    // merge in default options
+    for (var key in DEFAULT_OPTIONS) {
+        if (!opt.hasOwnProperty(key)) {
+            opt[key] = DEFAULT_OPTIONS[key];
+        }
+    }
+
+    /**
+     * a string of all allowed characters used for encoding/decoding.
+     * @type {String}
+     */
+    self.characters = opt.characters;
+
+    /**
+     * number of unique characters used to encode ids.
+     * @type {Number}
+     */
+    self.base = opt.base || opt.characters.length;
+
+    /**
+     * max length of id string.
+     * @type {Number}
+     */
+    self.length = opt.length || self._findLength(opt.max);
+}
+
+
+/**
+ * given an integer, find the maximum number of characters required to encode
+ * it given the number of characters we have available to work with (i.e.,
+ * base10 gives us 10 chars, base64 gives us 64 chars)
+ * @param {Number} max an integer
+ * @returns {Number}
+ */
+BaseN.prototype._findLength = function(max) {
```

Comment:
General comment, make sure to name your functions Foo.prototype.bar = function bar();

That way in core dumps you can actually find the function.